### DISCUSSION TITLE
Deleting events

### DISCUSSION THREAD

**Discussion by bobmanc** [vkurko/calendar#91] [Q&A] [answered]
I would like to have an X in the top right, to the right of ec-event-time, so a user can click it to delete the event. Is that possible? If not what is the best way to handle deleting events?

**Comment by bobmanc** [comment, marked as answer]
I figured it out in case anyone is wondering.

`	calendar.eventCalendar.setOption('eventContent', (info) =>
	{
		return {html: '<div class="cal-container"><div class="cal-header"><div>' + info.timeText + '</div><div>X</div></div><div class="cal-body">BODY</div></div>'};
	});
	calendar.eventCalendar.setOption('eventClick', (info) =>
	{
		if ((info.jsEvent.layerX > info.el.offsetWidth - 20) && (info.jsEvent.layerY < 20))
		{
			calendar.eventCalendar.removeEventById(info.event.id);
			setTimeout(() =>
			{
				calendar.saveCalendarEvents();
			}, 1000);
		}
	});
`

.cal-container {
	height: 100%;
	display: grid;
	grid-template-rows:20px 1fr;
}

.cal-header {
	display: flex;
	justify-content: space-between;
	align-items: center;
	padding: 5px;
}

.cal-body {
	display: flex;
	justify-content: center;
	align-items: center;
	font-size: 16px;
}

**Comment by vkurko** [comment]
One of the earlier examples is also available at the link in this [comment](https://github.com/vkurko/calendar/issues/25#issuecomment-942473353).

**Comment by bobmanc** [comment]
Thanks. Is there a way to know when an event is removed from the DOM. Right
now I wait 1 second to save it to the db.

**Reply by vkurko** [reply]
> Right now I wait 1 second to save it to the db.

Sorry, what's stopping you from sending a request right away?

**Comment by bobmanc** [comment]
If I send right away it doesn't pick up the change because, I assume, it
reads the DOM and the DOM has not finished updating.

**Reply by vkurko** [reply]
It's hard to understand how your code works, but something is clearly wrong with it if it depends on the DOM structure.

In general, the sequence of actions should be as follows:
1. The calendar loads events from the server (see [eventSources](https://github.com/vkurko/calendar#eventsources))
2. When you click on Delete, you send a delete request to the server
3. At the end of delete request, you call [refetchEvents()](https://github.com/vkurko/calendar#refetchevents) and the actions are repeated again

That is, the `removeEventById()` is not present at all in these actions.

**Comment by bobmanc** [comment]
In the demo you pointed to
JS Bin - Collaborative JavaScript Debugging
<https://jsbin.com/ximakozife/edit?js,output>

it says

eventClick: function(info) {
          if (info.event.display === 'auto') {
            let btn = info.el.querySelector('button');
            if (info.jsEvent.target === btn) {
              ec.removeEventById(info.event.id);
            }
          }
        }

**Reply by vkurko** [reply]
Yes, but this is a demo. It does not interact with the server and the database :smile:

**Comment by bobmanc** [comment]
Should I do something like this?

eventClick: function(info) {
          if (info.event.display === 'auto') {
            let btn = info.el.querySelector('button');
            if (info.jsEvent.target === btn) {
              ec  = ec.removeEventById(info.event.id);
             let events =  ec.getEvents();
// save events to the db here?
          }
          }
        }

**Comment by bobmanc** [comment]
That way I dont need to grab all the events from the db

**Comment by bobmanc** [comment]
I get what you're saying. I will relook at how I'm saving to the db.
Thanks

**Reply by vkurko** [reply]
No problem. Just remember that my recommendations are not the ultimate truth. Choose the method that best suits your case.
